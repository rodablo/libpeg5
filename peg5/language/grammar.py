from langkit.dsl import (
    T, ASTNode, Annotations, Field,  # Symbol,
    abstract, has_abstract_list, synthetic
)

from langkit.envs import (
    EnvSpec, add_to_env_kv,  # add_env
)

from langkit.expressions import (
    langkit_property, Self,  # ArrayLiteral, String, No, If, Property
)


@abstract
class Peg5Node(ASTNode):
    """
    Root class for Peg5 nodes.
    """
    #dummy = Property(Self.match(lambda j=T.Peg5: 2323))
    pass


@has_abstract_list
class Definition(Peg5Node):
    """
    Definition
    """
    id = Field(type=T.Identifier)
    expression = Field(type=T.Expression)
    #  doc = Field(type=T.CommentNode)
    env_spec = EnvSpec(
        add_to_env_kv(Self.id.sym, Self)
    )
    pass


class Definitions(Definition.list):

    annotations = Annotations(repr_name="Grammar")

    @langkit_property(return_type=T.Int, public=True)
    def n_dummy():
        """
        Return the number of definitions in this grammar
        """
        return Self.length


class Identifier(Peg5Node):
    """
    Identifier.
    """
    annotations = Annotations(repr_name="IdDef")

    token_node = True

    @langkit_property(return_type=T.Symbol, public=True)
    def sym():
        """
        Return the symbol for this identifier.
        """
        return Self.symbol


@abstract
@has_abstract_list
class AbstractPrimary(Peg5Node):
    """
    AbstractPrimary: Base for...
    TODO: suffix as boolean (remember to desugar the expression)
    """
    pass


@has_abstract_list
class SequenceList(AbstractPrimary.list):
    annotations = Annotations(repr_name="Primaries")


class ExpressionList(SequenceList.list):
    annotations = Annotations(repr_name="Choices")


class Expression(AbstractPrimary):
    """
    Expression
    """
    choices = Field(type=T.SequenceList.list)
    #choices = Field(type=T.ExpressionList)


@abstract
class AbstractOperator(Peg5Node):
    """
    AbstractOperator Node.
    """
    pass


class PrefixOperator(AbstractOperator):
    """
    """
    enum_node = True
    alternatives = ['and', 'not']


class PrefixedPrimary(AbstractPrimary):
    """
    PrefixedPrimary Node.
    """
    annotations = Annotations(repr_name="Prefixed")
    prefix = Field(type=T.PrefixOperator)
    primary = Field(type=T.AbstractPrimary)


class SuffixOperator(AbstractOperator):
    """
    """
    enum_node = True
    alternatives = ['optional', 'zero_or_more', 'one_or_more']


class SuffixedPrimary(AbstractPrimary):
    """
    SuffixedPrimary Node.
    """
    annotations = Annotations(repr_name="Suffixed")
    primary = Field(type=T.AbstractPrimary)
    suffix = Field(type=T.SuffixOperator)


class IdentifierReference(AbstractPrimary):
    """
    """
    annotations = Annotations(repr_name="IdRef")

    token_node = True


@abstract
class AbstractClass(AbstractPrimary):
    """
    Char Class Base.
    """
    token_node = True


class UnaryClass(AbstractClass):
    """
    UnaryClass
    """
    #a = Field(type=T.AbstractChar)
    pass


class BinaryClass(AbstractClass):
    """
    BinaryClass
    """
    #a = Field(type=T.AbstractChar)
    #b = Field(type=T.AbstractChar)
    pass


class Dot(AbstractPrimary):
    pass


class Literal(AbstractPrimary):
    """
    Literal
    """
    token_node = True


@synthetic
class CharNode(Peg5Node):
    """
    """
    # needs extension implementation
    @langkit_property(return_type=T.Character, external=True, public=True,
                      uses_entity_info=False, uses_envs=False)
    def denoted_value():
        """
        Return the value that this literal denotes.
        """
        pass
    #char = Field()
    pass


from peg5.language.lexer import (  # noqa: E402
    Token, peg5_lexer as L
)

# From libadalang/ada/grammar.py
# This import is after the language.ast import, because we want to be sure
# no class from langkit.expressions are shadowing the parser combinators.
from langkit.parsers import (  # noqa: E402
    Grammar, Or, List, Pick, Opt, NoBacktrack,
    #_, Skip,  # , NoBacktrack as cut, Null
)

peg5_grammar = Grammar(main_rule_name='definitions')
G = peg5_grammar

peg5_grammar.add_rules(

    # main_rule=List(Identifier(L.Identifier(match_text="first")),
    #               empty_valid=True),
    definitions=Pick(
        List(
            G.definition,
            list_cls=T.Definitions,
            empty_valid=True
        ),
        L.Termination
    ),

    definition=Definition(
        G.defining_id,
        G.expression,
        Opt(";")
    ),

    defining_id=Pick(
        G.identifier,
        NoBacktrack(),
        "<-"
    ),

    identifier=Identifier(
        Token.Identifier
    ),

    expression=Expression(
        G.choices
    ),

    choices=List(
        G.sequence,
        list_cls=T.ExpressionList,
        sep="/",
        empty_valid=False
    ),

    sequence=List(
        G.prefixed,
        list_cls=T.SequenceList,
        empty_valid=False
    ),

    prefixed=Or(
        PrefixedPrimary(
            Or(
                PrefixOperator.alt_and("&"),
                PrefixOperator.alt_not("!")
            ),
            G.suffixed
        ),
        G.suffixed
    ),

    suffixed=Or(
        SuffixedPrimary(
            G.primary,
            Or(
                SuffixOperator.alt_optional("?"),
                SuffixOperator.alt_zero_or_more("*"),
                SuffixOperator.alt_one_or_more("+")
            ),
        ),
        G.primary
    ),

    primary=Or(
        IdentifierReference(Token.Identifier),
        Pick(
            "(",
            NoBacktrack(),
            G.expression,
            ")"
        ),
        G.literal,
        G.char_class,
        Dot("."),
    ),

    char_class=Or(
        UnaryClass(L.UnaryRangeClass),
        BinaryClass(L.BinaryRangeClass),
    ),

    literal=Literal(Token.Literal),

    #char=CharNode(Token.Char),
)
