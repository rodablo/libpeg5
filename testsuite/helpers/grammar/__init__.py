import libpeg5lang as lp5l

_ctx = lp5l.AnalysisContext()


def do_parse_and_dump(source: str,
                      rule: str = lp5l.GrammarRule.main_rule_rule):

    unit = _ctx.get_from_buffer(filename='test.peg',
                                buffer=source,
                                rule=rule)
    print('==BEGIN==DUMP==')

    if unit.diagnostics:
        print('P&D - Found errors:')
        for d in unit.diagnostics:
            print('P&D - {}'.format(d))
    else:
        for t in unit.iter_tokens():
            print('{}'.format(t))
        print('===============')
        unit.root.dump()

    print('====END==DUMP==')
    return unit
