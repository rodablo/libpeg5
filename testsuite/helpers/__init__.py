import os
import sys


def print_done() -> None:
    print('Done.')


def run_make(additional_make_args=[]):
    """
    """
    from langkit.libmanage import DiagnosticError
    argv = ['make']
    argv.append('--build-mode=dev')
    #argv.append('--disable-all-mains')
    argv.append('--verbosity=none')
    #argv.append('--no-ada-api')
    #argv.append('--generate-unparser')
    argv.extend(additional_make_args)
    #
    from language import Manage
    m = Manage()

    return_code = m.run_no_exit(argv)
    # Flush stdout and stderr, so that diagnostics appear deterministically
    # before the script/program output.
    sys.stdout.flush()
    sys.stderr.flush()
    if return_code != 0:
        raise DiagnosticError()

    def update_os_environ(env_var: str, directory: str) -> None:
        path = os.environ.get(env_var)
        path = ('{}{}{}'.format(directory, os.path.pathsep, path) if path else env_var)
        os.environ[env_var] = path
        # print(env_var,' = ', path)

    m.setup_environment(update_os_environ)
