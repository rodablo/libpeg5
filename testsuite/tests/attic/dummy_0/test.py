from helpers.grammar import do_parse_and_dump

SRC = u"""
# some comment
aa <- bb

"""
unit = do_parse_and_dump(SRC)
