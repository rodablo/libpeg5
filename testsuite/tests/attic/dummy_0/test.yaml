driver: python
control:
- [SKIP, "True", "TODO: Not implemented."]