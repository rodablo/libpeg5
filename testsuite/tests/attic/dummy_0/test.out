==BEGIN==DUMP==
<Token NL '\n' at 1:1-2:1>
<Token Comment '# some comment' at 2:1-2:15>
<Token NL '\n' at 2:15-3:1>
<Token Identifier 'aa' at 3:1-3:3>
<Token Left_Arrow '<-' at 3:4-3:6>
<Token Identifier 'bb' at 3:7-3:9>
<Token NL '\n' at 3:9-4:1>
<Token NL '\n' at 4:1-5:1>
<Token Termination at 5:1-5:1>
===============
GrammarNode test.peg:1:1-5:1
|definitions:
|  DefinitionNodeList test.peg:1:1-5:1
|  |item_0:
|  |  DefinitionNode test.peg:3:1-5:1
|  |  |id:
|  |  |  Identifier test.peg:3:1-3:3: aa
|  |  |expression:
|  |  |  Expression test.peg:3:7-4:1
|  |  |  |choices:
|  |  |  |  SequenceList test.peg:3:7-4:1
|  |  |  |  |item_0:
|  |  |  |  |  Sequence test.peg:3:7-4:1
|  |  |  |  |  |primaries:
|  |  |  |  |  |  PrimaryNodeList test.peg:3:7-4:1
|  |  |  |  |  |  |item_0:
|  |  |  |  |  |  |  IdentifierReference test.peg:3:7-3:9: bb
====END==DUMP==